Feature - CLI Tutor Mode

jay-dee7 opened the pull request:
Adds the **CLI Tutor Mode**, which can help with learning the IPFS Command Line Interface from the Webui/Desktop. This feature was requested in #1421 

#### Done
* Option to enable/disable CLI Tutor Mode from the settings page
* If enabled, a `copy-icon` 📋 is shown next to the buttons/menus where those actions can be performed via IPFS' CLI
* This icon is visible at:
  * Files Page -> inside `+ Import` menu (Add File, Folder, Copy from IFPS)
  * Files Page -> File Menu (three dots next to a file) -> (Delete, Rename, Pin/Unpin, Download)
  * Peers Page -> next to `+ Add Connection`
  * Settings Page -> next to `Reset` button in config edit box
  * ~~Setting Page -> next to `Submit` button in API edit box~~

### Pending Work:
* ~~I recently discovered a state management issue with the modals, I couldn't find the root cause so far, it started happening after the refactor today (1st Aug 2020)~~
* ~~After rebasing from master today, there's a small UI glitch (text in the modal actions is left aligned, I'll work it out before making this PR reviewable)~~
* ~~I have added a bunch of TODO's in the code, which can potentially help with the code cleanup~~
* ~~Files component kind of looks like a mess right now, a lot of state is being shared across mulitple components (maybe we can refactor it to make it look neat)~~


Here's a visual for how it looks in action: [Jessica note: these are a bit outdated now since commits for visual tweaks on 12 Aug]

### Settings Page | Option to enable CLI Tutor Mode
![enable-ipfs-cli-tutor-mode](https://user-images.githubusercontent.com/10788442/89107652-261b0a80-d450-11ea-8d92-f5f49c6ca2f3.png)

### Files Page | CLI Tutor Mode Enabled
![cli-tutor-mode-files-page](https://user-images.githubusercontent.com/10788442/89107681-5c588a00-d450-11ea-82d2-ac3aae36c54e.png)

### Settings Page | CLI Tutor Mode - Open Modal
![cli-tutor-mode-open-modal](https://user-images.githubusercontent.com/10788442/89107669-4cd94100-d450-11ea-9e54-3a87e0d72186.png)



Signed-off-by: Jasdeep Singh <jasdeepsingh.uppal@gmail.com>

jay-dee7 commented:
@jessicaschilling here's the draft for cli tutor mode, the place which would need more attention is files component, since it has a lot of shared state, please check this out and suggest any refactors/changes for places where i might have broken any acceptable coding patterns. I'll update this PR tomorrow with little of code cleanup and modal actions alignment fix

jessicaschilling commented:
Thanks, @jay-dee7! I'd like @rafaelramalho19 to look through this in more detail ... but just a heads up that he's on holiday for the next week. Don't want you to think that we are ignoring you. 😊

jay-dee7 commented:
That's okay @jessicaschilling. In-fact apologies for the delay. This should've been up way before than it did. In the meantime I'll keep  on iterating  over this, just to see if anything can be improved upon 😄

jessicaschilling commented on src/components/cli-tutor-mode/CliTutorMode.js:
```diff
@@ -0,0 +1,109 @@
+  return (
+    <Modal {...props} className={className} onCancel={onLeave} style={{ maxWidth: '40em' }}>
+      <ModalBody icon={CopyIcon}>
+        <p className='charcoal w-80 center'>{t('copy the command to use it with ipfs command line interface')}</p>
```
```suggestion
        <p className='charcoal w-80 center'>{t('Copy the following and paste it into your terminal application to do this task in IPFS via the command line.')}</p>
```
Just a text edit.

jessicaschilling commented on src/components/cli-tutor-mode/CliTutorMode.js:
```diff
@@ -0,0 +1,109 @@
+      <ModalBody icon={CopyIcon}>
+        <p className='charcoal w-80 center'>{t('copy the command to use it with ipfs command line interface')}</p>
+        <div>
+          <Shell title="terminal shell">
```
```suggestion
          <Shell title="Shell">
```
Just "Shell" is fine here. Use the capital "S" in case we decide not to use CSS to uppercase these lines in future.

jay-dee7 commented:
sure @jessicaschilling i'll make these changes ASAP. There are more things that I didn't notice initially. ContextMenu in files pages needs some work, after rebasing, the copy icon and the text aren't aligned symmetrically. I'll try to get them fixed before the review starts

jay-dee7 commented:
@jessicaschilling I rebased from master just a while ago, one of the incoming changes were in `ApiAddressForm` component and since it was written using hooks (I believe), I was not able to overwrite props easily (I'm not sure what exactly broke the prop sharing, i believe its hooks but i have no idea.). So just for the `ApiAddressForm` component, I've managed the state for `cli-tutor-modal` locally to the component, instead of using the regular redux actions. Rest of the stuff stays almost same. I think this PR is ready for review 🤞

jessicaschilling commented:
Thanks, @jay-dee7! 
@rafaelramalho19 - are you able to take a look at this PR once you are back in and settled? Thank you!

rafaelramalho19 commented on src/components/cli-tutor-mode/CliTutorMode.js:
```diff
@@ -0,0 +1,110 @@
+  )
+}
+
+class CliTutorMode extends Component {
```
You're using a functional component above and here you used a class component. Can you make them equal? (I would advise choosing functional)

jay-dee7 replied:
I've Converted this into a functional component :)

rafaelramalho19 commented on src/components/cli-tutor-mode/CliTutorMode.js:
```diff
@@ -0,0 +1,110 @@
+          {t('cli.description')}
+        </p>
+        <p className='charcoal-muted w-90 center'>
+          {command.includes('ipfs config replace') ? t('cli.extraNotes') : ''}
```
Please change this to `command === cliCmdKeys.UPDATE_IPFS_CONFIG` (you need to import cliCmdKeys of course).

We should have a single source of truth for the command, just to avoid changing it on the constant files and breaking on this file.

jay-dee7 replied:
Thanks for digging this one, I've added comparison with the full command from `cliCmdKeys` now

rafaelramalho19 commented on src/components/cli-tutor-mode/CliTutorMode.js:
```diff
@@ -0,0 +1,110 @@
+        </div>
+        <div className='flex items-center'>
+          {
+            command && command.includes('ipfs config replace')
```
Replace here as suggested previously.

Also, why do you check for the existence of `command` here but not on line 25?

jay-dee7 replied:
added the same comparison here too

rafaelramalho19 commented on src/components/cli-tutor-mode/CliTutorMode.js:
```diff
@@ -0,0 +1,110 @@
+              style={{ height: '28px', transform: 'scale(1.5)', verticalAlign: 'bottom', color: 'dodgerblue' }}
+            /> : null
+          }
+          <Overlay show={isCliTutorModalOpen} onLeave={doOpenCliTutorModal}>
```
Just to be consistent shouldn't you do `() => doOpenCliTutorModal(false)`? I think that's the issue you had with being an object sometimes (I think that onLeave sends the click event as an argument)

jay-dee7 replied:
I think thats exactly what was causing the issue. I replaced the call with `() => doOpenCliTutorModal(false/true)` and it has been working just fine

rafaelramalho19 commented on src/components/cli-tutor-mode/CliTutorMode.js:
```diff
@@ -0,0 +1,110 @@
+CliTutorialModal.propTypes = {
+  onLeave: PropTypes.func.isRequired,
+  t: PropTypes.func.isRequired,
+  tReady: PropTypes.bool
```
You're not using tReady, why declare it?

jay-dee7 replied:
Removed this and added `command`, since command is a required field

rafaelramalho19 commented on src/files/FilesPage.js:
```diff
@@ -233,14 +236,19 @@ class FilesPage extends React.Component {
           onPin={() => this.props.doFilesPin(contextMenu.file.hash)}
-          onUnpin={() => this.props.doFilesUnpin(contextMenu.file.hash)} />
+          onUnpin={() => this.props.doFilesUnpin(contextMenu.file.hash)}
+          isCliTutorModeEnabled={isCliTutorModeEnabled} // @TODO connect ContextMenu to redux
```
Todos are actually done with `//TODO: `, and isn't it connected already?

jay-dee7 replied:
contextMenu page actually isn't connected but also it doesn't need to. Maybe if we'd want, we do can connect `Dropdown` component to reduce prop chaining

rafaelramalho19 commented on src/files/dropdown/Dropdown.js:
```diff
@@ -1,10 +1,22 @@
+      </button>
+      <button {...props} className={`bg-animate hover-bg-near-white pa2 pointer flex items-center  ${className}`}>
+        <CopyIcon {...props} onClick={onCliTutorMode} className='dib fill-current-color ph2 glow o-80 pointer'
+          style={{ height: '28px', transform: 'scale(1.3)', color: 'dodgerblue' }}
```
Try to avoid the transform (same as commented before)

jay-dee7 replied:
added a css class for this with `webkitTransform` but it would also work if we simple increased the height by 1.3X. Lemme know which way we'd like to move forward with :)

rafaelramalho19 commented on src/settings/SettingsPage.js:
```diff
@@ -56,6 +60,26 @@ export const SettingsPage = ({
+      <div className='mb4'>
+        <Title>{t('CLI TUTOR MODE')}</Title>
+        <Checkbox className='dib' onChange={doToggleCliTutorMode} checked={isCliTutorModeEnabled}
+          label={<span className='fw5 f6'>{t('Enable command-line interface (CLI) tutor mode')}</span>}/>
```
You need to add a translation key here, not the text

lidel replied:
Namely, add the key to `public/locales/en/settings.json` 
(no need to add it to other locales, that will happen automatically after PR is merged)

jay-dee7 replied:
@rafaelramalho19 @lidel i've replace these text lines with keys from translation files :)

rafaelramalho19 commented on src/settings/SettingsPage.js:
```diff
@@ -56,6 +60,26 @@ export const SettingsPage = ({
+          label={<span className='fw5 f6'>{t('Enable command-line interface (CLI) tutor mode')}</span>}/>
+        <div className='f6 charcoal lh-copy mw7'>
+          {
+            t('Enable this option to display')
```
Same here.

jay-dee7 replied:
replaced it

rafaelramalho19 commented on src/settings/SettingsPage.js:
```diff
@@ -56,6 +60,26 @@ export const SettingsPage = ({
+            className='dib fill-current-color ph2 glow o-80'
+            style={{ height: '28px', transform: 'scale(1.3)', verticalAlign: 'bottom', color: 'dodgerblue' }}/>
+          {
+            t(`next to common IPFS commands. Clicking an icon copies
```
And here

jay-dee7 replied:
and replaced this one too

jay-dee7 commented on src/bundles/cli-tutor-mode.js:
```diff
@@ -0,0 +1,54 @@
+export default {
+  name: 'cliTutorMode',
+  actionBaseType: 'CLI_TUTOR_MODE_TOGGLE',
+  init: store => {
```
@rafaelramalho19 this was bad on my part, this piece of code isn't required at all, I cleaned it up in my latest commit

jay-dee7 commented on src/bundles/cli-tutor-mode.js:
```diff
@@ -0,0 +1,54 @@
+  persist: true,
+  checkIfOnline: false,
+  doToggleCliTutorMode: key => ({ dispatch }) => {
+    const isEnabled = JSON.parse(localStorage.getItem('isCliTutorModeEnabled'))
```
@rafaelramalho19 I did add the suggested pattern however this was the first time I used redux-bundler. Please go through it once more just to validate if I did what was requested. I hope i didn't drift away too far :)

rafaelramalho19 replied:
You did good 👍

jay-dee7 commented on src/components/api-address-form/ApiAddressForm.js:
```diff
@@ -18,6 +23,22 @@ const ApiAddressForm = ({ t, doUpdateIpfsApiAddress, ipfsApiAddress = '' }) => {
+        <CopyIcon
+          onClick={setCliTutorModal}
+          className='dib fill-current-color ph2 glow o-80 pointer'
+          style={{ height: '28px', transform: 'scale(1.5)', verticalAlign: 'bottom', color: 'dodgerblue' }}
```
for this, i have now added two options, one using just the bigger pixel size (height is increased) and another one using `msTransform` and `webkitTransform`. I can quickly stick to one solution but as far as i can tell, they both work as expected. `webkitTransform` worked okay for me on safari too

rafaelramalho19 replied:
It works, the problem is that it renders a pixelated icon in safari (old issue unfortunately), so I would just stick with increasing the height

rafaelramalho19 commented:
After you remove the transforms, LGTM 😄

jay-dee7 commented:
thanks for the amazing review @rafaelramalho19, i've change the css class to use just height and no transforms :)

jessicaschilling commented:
@lidel -- do you mind going over this when you're able? I think we're very close to this being done.

NOTE: I want to get the "copy to clipboard" icon into ipfs-css before we ship. I can do that while you're reviewing 😊

Thanks!

jay-dee7 commented:
hey @jessicaschilling I just saw the changes that you made and I think they really addch the finishing touches to the feature, left aligned text in the modal looks great plus the new icon actually looks pretty amazing :) However, I did notice once thing though, the Checkbox on CLI Tutor Mode is slightly moved to bottom `lh-copy` class is doing the most of it. Can we do some sort of workaround for that?

I was going through the ipfs documentation and realised that a couple of commands are not correct I think, namely
* `Files Page -> + Import Button -> New Folder`
* `Files Page -> individual file -> three dots -> Rename`
I'll push the change for both of these commands just now but please let me know if rest of them are correct too

![image](https://user-images.githubusercontent.com/10788442/90262611-23ff7580-de6c-11ea-9465-cb3d2790427d.png)

lidel commented on src/bundles/files/consts.js:
```diff
@@ -40,3 +40,31 @@ export const DEFAULT_STATE = {
+  [cliCmdKeys.RENAME_IPFS_OBJECT]: (oldFileName, newFileName) => `ipfs files mv "${oldFileName}" ${newFileName}`,
+  [cliCmdKeys.ADD_FILE]: () => 'ipfs add <file-name>',
+  [cliCmdKeys.ADD_DIRECTORY]: () => 'ipfs add -r <folder-name>',
+  [cliCmdKeys.CREATE_NEW_DIRECTORY]: () => 'ipfs add -r <folder-name>',
```
```suggestion
  [cliCmdKeys.CREATE_NEW_DIRECTORY]: () => 'ipfs files mkdir <folder-name>',
```

jessicaschilling commented:
@jay-dee7 @lidel I'd vote for closing the dialog onclick of the "Copy" button. @jay-dee7, can you please change that?

jay-dee7 commented:
Sure @jessicaschilling I'll make the change ASAP :)

jessicaschilling commented:
@jay-dee7 And thanks for spotting the misaligned checkbox! Fixed it. 😊

jay-dee7 commented:
@jessicaschilling @lidel i've added both the requested points. 
Now by clicking the copy button in the modal, we copy the contents (command) and then close the modal.

Also, I've added a fix for rename command. I realized that we need to preserve the file path while renaming and we should wrap the full path path with quotes ("<path>"), because it'll throw errors if the full path contains spaces (for example a file named `cat image.png` would throw error in cli but if we wrap it with quotes as `"cat image.png"`) it won't. Please let me know if we're good to go with this :)

jessicaschilling commented:
@lidel - I'm happy but will wait for your OK, please feel free to merge 😊

lidel commented on tsconfig.json:
```diff
@@ -4,7 +4,7 @@
     "forceConsistentCasingInFileNames": true,
     "noImplicitReturns": true,
-    "noImplicitAny": true,
+    "noImplicitAny": false,
```
@Gozala how strongly do we  feel about this?
@jay-dee7 is it possible to add type info and keep this as `true`?
```suggestion
    "noImplicitAny": true,
```

jay-dee7 replied:
I just added type information for the methods where typescript transpiler was failing and reverted this option back to `true`. It works now as expected :)

lidel commented on src/files/modals/Modals.js:
```diff
@@ -123,15 +129,46 @@ class Modals extends React.Component {
+      case cliCmdKeys.RENAME_IPFS_OBJECT:
+        return cliCommandList[action](path, fileName)
+      case cliCmdKeys.PIN_OBJECT:
+        return cliCommandList[action](activeCid, isPinned ? 'rm' : 'add')
```
I believe WebUI pins recursively:
```suggestion
        return cliCommandList[action](activeCid, isPinned ? 'rm' : 'add -r')
```

jay-dee7 replied:
@lidel by recursive you mean that if we pin a directory, all the files in the directory and any of it's sub-directories and their files gets pinned?

jay-dee7 replied:
i just confirmed, webui doesn;t add/remove pin recursively
![ipfs-add-pin](https://user-images.githubusercontent.com/10788442/90814073-2c612000-e346-11ea-840f-807d2ffc8f9e.png)
 but please let me know which approach should i follow :)

lidel reviewed (changes requested):
@jay-dee7 nearly there! small asks inline :pray:

jay-dee7 commented on src/bundles/files/consts.js:
```diff
@@ -40,3 +40,34 @@ export const DEFAULT_STATE = {
+
+export const cliCommandList = {
+  [cliCmdKeys.UPDATE_IPFS_CONFIG]: () => 'ipfs config replace <path-to-settings.json>',
+  [cliCmdKeys.DELETE_FILE_FROM_IPFS]: (cid) => `ipfs files rm ${cid}`,
```
@lidel thanks for catching this one. I'll change it to use full filepath and also recursive. I think 
```ipfs files rm -r ${filepath}```
is a much safer option (command would succeed in more cases regardless of a directory/file)

lidel reviewed (approved):
Thank you @jessicaschilling, I think the old i18next issue is gone, all looks good.

----

I've made a small fix in 5641184 to ensure CLI tutor lists import actions as the button in GUI 
(`ipfs add && ipfs files cp`, context: https://github.com/ipfs-shipyard/ipfs-webui/issues/676)

Ideally `/<dest-name>` in import commands would be replaced with `"${path}/<dest-name>"` where `path` is the current dir in MFS. I started doing that in 3a337de but had no bandwidth to wire it up.

I think it's ok to merge with static destination, but if anyone has time to ensure `path` is set for [these actions](https://github.com/jay-dee7/ipfs-webui/blob/jay-dee7/cli-tutor-mode/src/files/modals/Modals.js#L157-L160) that would be even better for user experience.

jessicaschilling commented:
@lidel -- Thanks for calling the subdirectory detail out. I'm inclined to merge this as-is and address that in a separate issue (https://github.com/ipfs-shipyard/ipfs-webui/issues/1601), since I suspect we may want to make some tweaks to tutor mode overall after this gets into the hands of users for a bit.

I've also opened https://github.com/ipfs-shipyard/ipfs-webui/issues/1600 for consolidating repeat-use i18n keys, which I can probably take care of in the next week or two.

Thanks all (especially @jay-dee7!) for all of the work!
